src/metabase/upload.clj

```diff
@@ -797,33 +798,36 @@
         (ex-info (tru "You don''t have permissions to do that.")
                  {:status-code 403}))))
 
-(defn- check-can-append
+(defn- check-can-update
   "Throws an error if the user cannot upload to the given database and schema."
   [db table]
-  (when-let [error (can-append-error db table)]
+  (when-let [error (can-update-error db table)]
     (throw error)))
 
 (defn can-upload-to-table?
   "Returns true if the user can upload to the given database and table, and false otherwise."
   [db table]
-  (nil? (can-append-error db table)))
+  (nil? (can-update-error db table)))
 
 ;;; +--------------------------------------------------
-;;; |  public interface for appending to uploaded table
+;;; |  public interface for updating an uploaded table
 ;;; +--------------------------------------------------
 
-(mu/defn append-csv!
-  "Main entry point for appending to uploaded tables with a CSV file.
+(mu/defn update-csv!
+  "Main entry point for updating an uploaded table with a CSV file.
   This will create an auto-incrementing primary key (auto-pk) column in the table for drivers that supported uploads
-  before auto-pk columns were introduced by metabase#36249."
-  [{:keys [^File file table-id]}
+  before auto-pk columns were introduced by metabase#36249, if it does not already exist."
+  ;; TODO: update callers to all pass an explicit action, and make it required
+  [{:keys [^File file table-id action] :or {action ::append}}
    :- [:map
        [:table-id ms/PositiveInt]
-       [:file (ms/InstanceOfClass File)]]]
+       [:file (ms/InstanceOfClass File)]
+       [:action keyword?]]]
   (let [table    (api/check-404 (t2/select-one :model/Table :id table-id))
-        database (table/database table)]
-    (check-can-append database table)
-    (append-csv!* database table file)))
+        database (table/database table)
+        replace? (= ::replace action)]
```

Comment:
Big difference imho between `(foo bar baz quux true)` and `(foo bar baz quux :replace? true)`, but I know what you mean. +1 to getting the Malli schema out sooner rather than later then.

And yeah, even more reasonable in light of `::merge`

Reply:
I agree that keyword arguments solve the problem. Some of the bias comes languages without kwargs. Isn't `(foo bar baz quux ::replace)` even more snappy though 😋[^1]

[^1]: Yeah.. I do swap out for a `:replace-rows?` kwarg lower down the stack, but that's only because I don't want logic about the action accumulating in the innards. The "business logic" turns the action-noun into instruction. Trivial now but kinda speculating that `::merge` might involve teasing out a few options.

Added a schema for them and using it in both the API layers and the internals.

Reply:
Oh last remark, the `:replace-rows?` thing is also totally a smell. `update-with-csv!` is a very smelly method - too many arguments, too long, too many levels of abstraction etc 😄